File: solutions/csharp/squeaky-clean/1/SqueakyClean.cs
using System;
using System.Text;
using System.Text.RegularExpressions;

public static class Identifier
{
    public static string Clean(string identifier)
    {
        var sb = new StringBuilder(identifier, 100);
        sb.Replace(' ', '_');
        
        var str = sb.ToString();
        str = string.Join("", str.ToCharArray().Select(c => Char.IsControl(c) ? "CTRL" : c.ToString()));

        var camelChars = new List<char>();
        var prevDash = false;
        foreach (char ch in str.ToCharArray())
        {
            if (ch == '-')
            {
                prevDash = true;
            }
            else
            {
                if (prevDash)
                {
                    camelChars.Add(Char.ToUpper(ch));
                }
                else
                {
                    camelChars.Add(ch);
                }
                prevDash = false;
            }
        }
        str = new string(camelChars.ToArray());

        str = Regex.Replace(str, @"[^\w]", "");
        str = Regex.Replace(str, @"[\d]", "");
        str = Regex.Replace(str, @"[α-ω]", "");

        return str;
    }
}

File: solutions/java/squeaky-clean/1/src/main/java/SqueakyClean.java
class SqueakyClean {
    static String clean(String identifier) {
        var sb = new StringBuilder();
        var isKabab = false;
        for (var ch: identifier.toCharArray()) {
            if (Character.isWhitespace(ch)){
                sb.append('_');
            } else if (ch == '-') {
                isKabab = true;  
            } else if (isKabab && Character.isLetter(ch)) {
                sb.append(Character.toUpperCase(ch));
                isKabab = false;
            } else if (Character.isDigit(ch)) {
                if (ch == '4') {
                    sb.append('a');
                } else if (ch == '3') {
                    sb.append('e');
                } else if (ch == '0') {
                    sb.append('o');
                } else if (ch == '1') {
                    sb.append('l');
                } else if (ch == '7') {
                    sb.append('t');
                } else {
                    sb.append(ch);
                }
            } else if (Character.isLetter(ch)) {
                sb.append(ch);
            }
        }
        return sb.toString();
    }
}
